### PR TITLE
[#9536] Student home E2E test migration

### PR DESCRIPTION
**Pull request by jacoblipech**
part of #9536 

- Copied the files over from for `StudentHomePageE2ETest` to work
- Added a `databundle` API endpoint to handle requests send from E2E test to work
- Fix E2E tests for `StudentHomePageE2ETest`

### CODE REVIEW

File: src/e2e/java/teammates/e2e/pageobjects/AppPageNew.java [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+    public void clickLoginDropDown() {
+        click(By.cssSelector("a.nav-link.dropdown-toggle"));
+    }
+
+    public boolean verifyStrangerMessage() {
+        return browser.driver.findElement(By.tagName("h5")).getText()
+                .contains("Your Google account is not known to TEAMMATES");
+    }
+
+    public boolean verifyFirstPanelFeedbackSessionName(String message) {
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/AppPageNew.java]: This is specific to student home page, so it should not be in this parent class.

File: src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        AppUrl detailsPageUrl = createUrl(Const.WebPageURIs.STUDENT_HOME_PAGE)
+                .withUserId(testData.students.get("SHomeUiT.charlie.d@SHomeUiT.CS2104").googleId);
+
+        StudentHomePageNew studentHomePage = loginAdminToPageNew(detailsPageUrl, StudentHomePageNew.class);
+        browser.waitForPageLoad();
+
+        assertTrue(studentHomePage.verifyAdminPagePresent());
+    }
+
+    private void testLinks() {
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java]: No need to test this anymore. The links are already tested as part of front-end unit test.

File: src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+        logout();
+        studentHome = getHomePageNew().clickStudentLogin()
+                                   .loginAsStudent(TestProperties.TEST_STUDENT1_ACCOUNT,
+                                                   TestProperties.TEST_STUDENT1_PASSWORD);
+
+        browser.waitForPageLoad();
+        assertTrue(studentHome.verifyFirstPanelFeedbackSessionName("[SHomeUiT.CS1101]: Programming Methodology"));
+
+        ______TS("login successfully as admin (masquerade mode)");
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java]: Similarly, no need to test this anymore. This is not specific to student home page.

File: src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+        ______TS("login successfully");
+
+        logout();
+        studentHome = getHomePageNew().clickStudentLogin()
+                                   .loginAsStudent(TestProperties.TEST_STUDENT1_ACCOUNT,
+                                                   TestProperties.TEST_STUDENT1_PASSWORD);
+
+        browser.waitForPageLoad();
+        assertTrue(studentHome.verifyFirstPanelFeedbackSessionName("[SHomeUiT.CS1101]: Programming Methodology"));
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java]: Instead of `verifyFirstPanel`, why not pass a list of course id+names and have everything verified?

While you're at it, check the list of feedback sessions for all courses as well.

File: src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        studentHomePageNew.clickSubmitFeedbackButton("First Feedback Session");
+        studentHomePageNew.waitForPageToLoad();
+
+    }
+
+    @AfterClass
+    public void classTearDown() {
+        // The courses of test student1 account is not the same as `StudentProfilePageUiTest`
+        // so it has to be cleared before running that test. This is the reason why `StudentProfilePageUiTest`
+        // would fail if we don't remove the data bundle in this test.
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java]: This is no longer (or not yet) the case now, so the block has no reason to be here (yet). Let's remove it.

File: src/e2e/java/teammates/e2e/pageobjects/HomePageNew.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    @Override
+    protected boolean containsExpectedPageContents() {
+        return getTitle().contains("TEAMMATES");
+    }
+
+    public LoginPageNew clickStudentLogin() {
+        click(studentLoginLink);
+        waitForElementVisibility(By.tagName("h3"));
```
[the remark this answers is not in the document]
Comment (jacoblipech) [reply src/e2e/java/teammates/e2e/pageobjects/HomePageNew.java]: this is the tag for the `Not logged in` in the login page which applies to both student and instructorr

File: src/e2e/java/teammates/e2e/pageobjects/AppPageNew.java (hunk 1 above)
Comment (jacoblipech) [reply src/e2e/java/teammates/e2e/pageobjects/AppPageNew.java]: I was thinking in the future can be applied to instructor home page too since there is panel involved in the same format?

Comment (wkurniawan07) [reply src/e2e/java/teammates/e2e/pageobjects/AppPageNew.java]: Then you're adding a coupling between those two pages.

Comment (jacoblipech) [reply src/e2e/java/teammates/e2e/pageobjects/AppPageNew.java]: oh ya, my bad! I forgot that they are not sharing the same panel component in the front end.

File: src/web/app/page.component.html [hunk 7]
```diff
@@ -107,7 +109,6 @@ <h5 style="text-align: center;"><i class="fas fa-exclamation-circle"></i>&nbsp;O
       </div>
     </div>
   </div>
-  <h1 *ngIf="pageTitle && isValidUser">{{ pageTitle }}</h1>
```
Comment (wkurniawan07) [inline src/web/app/page.component.html]: Why is this moved up and why the `isValidUser` condition check is removed?

File: src/web/app/page.component.ts [hunk 8]
```diff
@@ -75,6 +75,9 @@ export class PageComponent implements OnInit {
         });
       }
     });
+    if (environment.frontendUrl) {
+      this.logoutUrl += `?frontendUrl=${environment.frontendUrl}`;
+    }
```
Comment (wkurniawan07) [inline src/web/app/page.component.ts]: Don't think you need this anymore?

File: src/web/app/page.component.ts [hunk 9]
```diff
@@ -42,6 +41,7 @@ export class PageComponent implements OnInit {
   messageList: StatusMessage[] = [];
   isNetworkOnline$: Observable<boolean>;
   version: string = environment.version;
+  logoutUrl: string = `${environment.backendUrl}/logout`;
```
Comment (wkurniawan07) [inline src/web/app/page.component.ts]: Don't think you need this anymore?

File: src/web/app/page.component.html [hunk 10]
```diff
@@ -35,12 +35,13 @@
           <a class="dropdown-item" routerLink="/web/instructor" *ngIf="isInstructor">To instructor pages</a>
           <a class="dropdown-item" routerLink="/web/admin" *ngIf="isAdmin">To admin pages</a>
           <div class="dropdown-divider" *ngIf="isStudent || isInstructor || isAdmin"></div>
-          <a class="dropdown-item" [href]="logoutUrl">Log Out</a>
+          <a class="dropdown-item" id="btnLogout" [href]="logoutUrl">Log Out</a>
```
Comment (wkurniawan07) [inline src/web/app/page.component.html]: Let's use `kebab-case` for HTML IDs.

File: src/main/java/teammates/common/util/Const.java [hunk 11]
```diff
@@ -562,6 +562,8 @@ private Const() {
         public static final String IS_STUDENT_REJOINING = "isstudentrejoining";
         public static final String IS_INSTRUCTOR_REJOINING = "isinstructorrejoining";
 
+        public static final String BACKDOOR_DATA = "data";
+
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Const.java]: Not used anymore?

File: src/e2e/java/teammates/e2e/pageobjects/LoginPage.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
+package teammates.e2e.pageobjects;
+
+/**
+ * An abstract class of login page for the app to provide ways to interact validate some aspects of it.
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/LoginPage.java]: Abstract Page Object Model for ...

File: src/e2e/java/teammates/e2e/pageobjects/HomePage.java [hunk 13]
```diff
@@ -0,0 +1,28 @@
+package teammates.e2e.pageobjects;
+
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+/**
+ * Represents the home page of the website.
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/HomePage.java]: Page Object Model for ...

File: src/e2e/java/teammates/e2e/pageobjects/GoogleLoginPage.java [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+
+import com.google.common.base.Preconditions;
+
+import teammates.e2e.util.TestProperties;
+
+/**
+ * Page Object Model for login page using official Google account.
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/GoogleLoginPage.java]: It's not a "login page using official Google account". It's the "official Google Accounts login page".

File: src/e2e/java/teammates/e2e/pageobjects/AppPage.java [hunk 15]
```diff
@@ -0,0 +1,621 @@
...
+
+    /**
+     * Used by subclasses to create a {@code AppPage} object to wrap around the
+     * given {@code browser} object. Fails if the page content does not match
+     * the page type, as defined by the sub-class.
+     */
+    public AppPage(Browser browser) {
+        this.browser = browser;
+        this.firefoxChangeHandler = new FirefoxChangeHandler(); //legit firefox
+        // jQueryAjaxHandler = new JQueryAjaxHandler(); // no longer used
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/AppPage.java]: Can just remove this line of code.

File: src/e2e/java/teammates/e2e/pageobjects/AppPage.java [hunk 16]
```diff
@@ -0,0 +1,621 @@
...
+    public <E> E waitFor(ExpectedCondition<E> expectedCondition) {
+        WebDriverWait wait = new WebDriverWait(browser.driver, TestProperties.TEST_TIMEOUT);
+        return wait.until(expectedCondition);
+    }
+
+    /**
+     * Waits until the page is fully loaded.
+     */
+    public void waitForPageToLoad() {
+        waitFor(driver -> {
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/AppPage.java]: This can be replaced with `browser.waitForPageToLoad()`.

File: src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+            String feedbackSessionNames = testData.feedbackSessions.entrySet().stream()
+                    .filter(feedbackSession -> courseId.equals(feedbackSession.getValue().getCourseId()))
+                    .map(x -> x.getValue().getFeedbackSessionName())
+                    .collect(Collectors.joining());
+
+            if (feedbackSessionNames.isEmpty()) {
+                assertFalse(verifyVisibleFeedbackSessionToStudents("empty", i));
+            } else {
+                assertTrue(verifyVisibleFeedbackSessionToStudents(feedbackSessionNames, i));
+            }
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java]: I'm not sure of what this entire block is doing.
- `feedbackSessionNames` or `feedbackSessionName`? Seems like you are testing one thing only.
- If there is no feedback session, what are you trying to assert there?

File: src/e2e/java/teammates/e2e/pageobjects/AppPage.java [hunk 18]
```diff
@@ -0,0 +1,621 @@
...
+
+        // Fail safe in case the implementation of staleness checks is changed
+        if (isExpectedCondition(ExpectedConditions.stalenessOf(element))) {
+            throw new AssertionError(
+                    "Element is stale but should have been caught earlier by element.isEnabled().");
+        }
+
+        @SuppressWarnings("unchecked")
+        Map<String, Object> result = (Map<String, Object>) executeScript(
+                "const element = arguments[0];"
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/AppPage.java]: For long (> 3 lines) scripts like this (not just here), let's extract them out. You can check here for example:

https://github.com/TEAMMATES/teammates/blob/f966efc68b19a39b58dccc5d48dccea397996b76/src/e2e/resources/scripts/waitForPageLoad.js#L1-L26
https://github.com/TEAMMATES/teammates/blob/f966efc68b19a39b58dccc5d48dccea397996b76/src/e2e/java/teammates/e2e/pageobjects/Browser.java#L24-L33
https://github.com/TEAMMATES/teammates/blob/f966efc68b19a39b58dccc5d48dccea397996b76/src/e2e/java/teammates/e2e/pageobjects/Browser.java#L80-L85

File: src/web/app/page.component.html (hunk 7 above)
Comment (jacoblipech) [reply src/web/app/page.component.html]: This was originally moved up to show check the pageTitle in `containsExpectedPageContents()` despite showing error message in E2E test. Since that test is not required, I will remove this as well

File: src/web/app/page.component.ts (hunk 8 above)
Comment (jacoblipech) [reply src/web/app/page.component.ts]: Sorry these were added due to the rebase I think. Will remove them.

File: src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java (hunk 17 above)
Comment (jacoblipech) [reply src/e2e/java/teammates/e2e/cases/e2e/StudentHomePageE2ETest.java]: My intention for this feedbackSessionName is that:
- if the `feedbackSessionName` is empty, the tag found in html should be empty as well.

Actually I can simply `assertTrue(feedbackSessionName, i)` as the condition should return true given an empty string or an actual feedbackSessionName.

File: src/web/app/page.component.ts (hunk 8 above)
Comment (wkurniawan07) [reply src/web/app/page.component.ts]: Uh-oh, this should not have been removed.

File: src/web/app/page.component.ts (hunk 9 above)
Comment (wkurniawan07) [reply src/web/app/page.component.ts]: This one also should not have been removed.
